src/gui/editors/PianoRoll.cpp

```diff
@@ -2550,10 +2566,9 @@ void PianoRoll::dragNotes( int x, int y, bool alt, bool shift, bool ctrl )
 
 
 	// if they're not holding alt, quantize the offset
-	if( ! alt )
+	if (!alt)
```

Comment:
Why not move this so off_ticks is only quantized if you're in nudge mode? Then you should be able to use off_ticks for the snap calculations as well, instead of recalculating it from mouse positions and stuff below.

Reply:
I'm not sure we can really reuse `off_ticks` for the snap mode: `off_ticks` is the offset from the initial position the mouse clicked in ticks, while the position we calculate in the snap mode is the absolute mouse position in ticks on the time line. Maybe if we added `m_moveStartX * TimePos::ticksPerBar() / m_ppb;` to it, but then we are having to do the conversion to ticks again anyways, right?

Reply:
`note->oldPos().getTicks() + off_ticks` should give the absolute position, no?

Reply:
In fact, this is exactly what the existing code already does at line 2598 (2645 in the PR version at time of writing).

```
int pos_ticks = note->oldPos().getTicks() + off_ticks;
...
note->setPos(TimePos(pos_ticks));
```

Reply:
Sure, I'm giving it a shot. Though I still think this will become a rabbit hole and I'll end up rewriting more than expected.

Reply:
Last two commits separate the nudge and snap logic more explicitly and reuse off_ticks as you suggested. It required some rearrangement of code. Method still needs refactoring and I tried to rewrite the least possible. A bug that is present on `master` was fixed on the calculation of the boundaries limits for the note keys.